### PR TITLE
Make VM async

### PR DESCRIPTION
**Pull request by tballmsft**

### CODE REVIEW

File: src/test/testrunner.ts [hunk 1]
```diff
@@ -643,7 +643,7 @@ export class JDServiceTestRunner extends JDServiceClient {
     }
 
     public refreshEnvironment() {
```
Comment (pelikhan) [inline src/test/testrunner.ts:645]: this should be async too

File: src/test/testrunner.ts [hunk 2]
```diff
@@ -643,7 +643,7 @@ export class JDServiceTestRunner extends JDServiceClient {
...
-        this._env.refreshEnvironment();
+        this._env.refreshRegistersAsync();
```
Comment (pelikhan) [inline src/test/testrunner.ts:646]: await

File: src/test/testrunner.ts [hunk 3]
```diff
@@ -653,7 +653,7 @@ export class JDServiceTestRunner extends JDServiceClient {
 
     // eslint-disable-next-line @typescript-eslint/no-explicit-any
     public writeRegister(regName: string, val: any) {
```
Comment (pelikhan) [inline src/test/testrunner.ts:655]: async

File: src/test/testrunner.ts [hunk 4]
```diff
@@ -653,7 +653,7 @@ export class JDServiceTestRunner extends JDServiceClient {
...
-        return this._env.writeRegister(regName, val)
+        return this._env.writeRegisterAsync(regName, val)
```
Comment (pelikhan) [inline src/test/testrunner.ts:656]: await

File: src/vm/environment.ts [hunk 5]
```diff
@@ -174,16 +157,16 @@ export class VMEnvironment extends JDEventSource {
         return this._envs[root]
     }
 
-    public refreshEnvironment() {
-        Object.values(this._envs).forEach(s => s?.refreshEnvironment())
+    public async refreshRegistersAsync() {
+        Object.values(this._envs).forEach(async s => {
```
Comment (pelikhan) [inline src/vm/environment.ts:161]: your leaking the promises. do
```
for(const s of Object.values(this._envs)) {
   await s?.refresh...
}

File: src/vm/vmrunner.ts [hunk 6]
```diff
@@ -120,7 +120,7 @@ class IT4CommandEvaluator {
                 const reg = args[0] as jsep.MemberExpression
                 if (
                     (this.inst === "writeRegister" &&
-                        this.env.writeRegister(reg, ev)) ||
+                        this.env.writeRegisterAsync(reg, ev)) ||
```
Comment (pelikhan) [inline src/vm/vmrunner.ts:123]: this will always be true since it returns a Promise

File: src/vm/vmrunner.ts [hunk 7]
```diff
@@ -314,28 +314,28 @@ export class IT4ProgramRunner extends JDEventSource {
...
+    async run() {
         if (!this._running) return
         try {
-            this._env.refreshEnvironment()
+            await this._env.refreshRegistersAsync()
             if (this._waitQueue.length > 0) {
                 const nextTime: IT4HandlerRunner[] = []
-                this._waitQueue.forEach(h => {
-                    h.step()
+                this._waitQueue.forEach(async h => {
```
Comment (pelikhan) [inline src/vm/vmrunner.ts:337]: leaking promises

**Review by pelikhan — CHANGES_REQUESTED** [review]
A few places missed.

File: src/vm/environment.ts [hunk 8]
```diff
@@ -8,15 +8,9 @@ import { JDEventSource } from "../jdom/eventsource"
...
-export async function refreshRegistersAsync(registers: SMap<JDRegister>) {
-    for (const k in registers) {
-        const register = registers[k]
-        let retry = 0
-        let val: any = undefined
-        do {
-            await register.refresh()
-            val = register.unpackedValue?.[0]
-        } while (val === undefined && retry++ < 2)
+export class VMError {
```
Comment (pelikhan) [inline src/vm/environment.ts:11]: use this pattern: https://github.com/microsoft/jacdac-ts/blob/main/src/jdom/error.ts

File: src/vm/environment.ts [hunk 9]
```diff
@@ -103,7 +101,15 @@ export class VMServiceEnvironment extends JDServiceClient {
     }
 
     public async refreshRegistersAsync() {
-        await refreshRegistersAsync(this._registers)
+        for (const k in this._registers) {
+            const register = this._registers[k]
+            let retry = 0
+            let val: any = undefined
+            do {
+                await register.refresh()
```
Comment (pelikhan) [inline src/vm/environment.ts:109]: refresh does retries already

File: src/vm/expr.ts [hunk 10]
```diff
@@ -62,6 +46,12 @@ export function unparse(e: jsep.Expression): string {
     }
 }
 
+export class ExprEvaluatorException {
```
Comment (pelikhan) [inline src/vm/expr.ts:49]: use a single exception type with various error codes

File: src/test/testrunner.ts [hunk 11]
```diff
@@ -364,10 +364,9 @@ class JDCommandEvaluator {
                 )
                 const ev = expr.eval(args[1])
                 const reg = args[0] as jsep.Identifier
-                if (this.testRunner.serviceTestRunner.writeRegister(reg.name, ev)) {
-                    this._status = JDTestCommandStatus.Passed
-                    this._progress = `wrote ${ev} to register ${reg.name}`
-                }
+                await this.testRunner.serviceTestRunner.writeRegisterAsync(reg.name, ev)
+                this._status = JDTestCommandStatus.Passed
```
Comment (pelikhan) [inline src/test/testrunner.ts:368]: you should use a property setter and automatically emit a change event when status change
same for progress.

File: src/vm/environment.ts (hunk 9 above)
Comment (tballmsft) [reply src/vm/environment.ts:106]: you wrote this code for me. Are you sure?

**Comment by pelikhan** [comment]
Yeah let’s simplify

________________________________

**Pull request merged**